storage/innobase/xtrabackup/src/backup_mysql.cc

```diff
@@ -983,14 +987,46 @@ stop_query_killer()
 	os_event_wait_time(kill_query_thread_stopped, 60000);
 }
 
+
+/*********************************************************************//**
+Function acquires a backup tables lock if supported by the server.
+Allows to specify timeout in seconds for attempts to acquire the lock.
+@returns true if lock acquired */
+bool
+lock_tables_for_backup(MYSQL *connection, int timeout)
+{
+	if (have_lock_wait_timeout) {
+		char query[200];
+
+		ut_snprintf(query, sizeof(query),
+			    "SET SESSION lock_wait_timeout=%d", timeout);
+
+		xb_mysql_query(connection, query, false);
+	}
+
+	if (have_backup_locks) {
+		msg_ts("Executing LOCK TABLES FOR BACKUP...\n");
+		xb_mysql_query(connection, "LOCK TABLES FOR BACKUP", true);
+		tables_locked = true;
+		return(true);
+	}
+
+	msg_ts("Error: LOCK TABLES FOR BACKUP is not supported.\n");
+
+	return(false);
+}
+
 /*********************************************************************//**
 Function acquires either a backup tables lock, if supported
 by the server, or a global read lock (FLUSH TABLES WITH READ LOCK)
 otherwise.
 @returns true if lock acquired */
 bool
-lock_tables(MYSQL *connection)
+lock_tables_maybe(MYSQL *connection)
 {
+	if (tables_locked || opt_lock_ddl_per_table) {
+		return(true);
+	}
 	if (have_lock_wait_timeout) {
 		/* Set the maximum supported session value for
 		lock_wait_timeout to prevent unnecessary timeouts when the
```

Comment:
I guess, line below might be a good place to utilize that new `lock_tables_for_backup` function.